File: Assets/Scripts/GameManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

/// <summary>
/// Main game controller that manages the overall game state and progression
/// </summary>
public class GameManager : MonoBehaviour
{
    public enum GameState
    {
        Welcome,
        CharacterSelection,
        TribeDivision,
        Camp,
        Challenge,
        TribalCouncil,
        Merge,
        FinalTribalCouncil,
        GameOver
    }

    public enum GamePhase
    {
        PreMerge,
        PostMerge,
        Final
    }

    // Singleton instance
    public static GameManager Instance { get; private set; }

    // Current game state
    [SerializeField] private GameState _currentState;
    [SerializeField] private GamePhase _currentPhase;
    
    // Game data
    [SerializeField] private int _currentDay = 1;
    [SerializeField] private int _survivorsRemaining = 20;
    [SerializeField] private int _mergeThreshold = 12;
    [SerializeField] private int _finalThreshold = 3;
    [SerializeField] private List<SurvivorData> _allSurvivors = new List<SurvivorData>();
    [SerializeField] private List<TribeData> _tribes = new List<TribeData>();
    [SerializeField] private List<SurvivorData> _jury = new List<SurvivorData>();
    [SerializeField] private SurvivorData _playerSurvivor;
    [SerializeField] private TribeData _playerTribe;
    [SerializeField] private TribeData _mergedTribe;

    // References to systems
    public RelationshipSystem RelationshipSystem { get; private set; }
    public AllianceSystem AllianceSystem { get; private set; }
    public ChallengeSystem ChallengeSystem { get; private set; }
    public TribalCouncilSystem TribalCouncilSystem { get; private set; }
    public EnergySystem EnergySystem { get; private set; }
    public IdolSystem IdolSystem { get; private set; }
    public UIManager UIManager { get; private set; }

    private void Awake()
    {
        // Singleton pattern implementation
        if (Instance == null)
        {
            Instance = this;
            DontDestroyOnLoad(gameObject);
            InitializeGame();
        }
        else
        {
            Destroy(gameObject);
        }
    }

    private void InitializeGame()
    {
        // Initialize all game systems
        RelationshipSystem = GetComponentInChildren<RelationshipSystem>() ?? gameObject.AddComponent<RelationshipSystem>();
        AllianceSystem = GetComponentInChildren<AllianceSystem>() ?? gameObject.AddComponent<AllianceSystem>();
        ChallengeSystem = GetComponentInChildren<ChallengeSystem>() ?? gameObject.AddComponent<ChallengeSystem>();
        TribalCouncilSystem = GetComponentInChildren<TribalCouncilSystem>() ?? gameObject.AddComponent<TribalCouncilSystem>();
        EnergySystem = GetComponentInChildren<EnergySystem>() ?? gameObject.AddComponent<EnergySystem>();
        IdolSystem = GetComponentInChildren<IdolSystem>() ?? gameObject.AddComponent<IdolSystem>();
        UIManager = GetComponentInChildren<UIManager>() ?? gameObject.AddComponent<UIManager>();

        // Load survivor database
        LoadSurvivorDatabase();
        
        // Set initial game state
        SetGameState(GameState.Welcome);
        _currentPhase = GamePhase.PreMerge;
    }

    private void LoadSurvivorDatabase()
    {
        // Load survivor data from JSON file
        TextAsset survivorJson = Resources.Load<TextAsset>("Data/SurvivorDatabase");
        if (survivorJson != null)
        {
            SurvivorDatabase database = JsonUtility.FromJson<SurvivorDatabase>(survivorJson.text);
            _allSurvivors = database.survivors;
        }
        else
        {
            Debug.LogError("Failed to load survivor database!");
        }
    }

    public void SetGameState(GameState newState)
    {
        _currentState = newState;
        UIManager.UpdateUI(newState);
        
        // Handle state specific logic
        switch (newState)
        {
            case GameState.Welcome:
                break;
                
            case GameState.CharacterSelection:
                break;
                
            case GameState.TribeDivision:
                DivideTribes();
                break;
                
            case GameState.Camp:
                EnergySystem.RefillEnergy();
                break;
                
            case GameState.Challenge:
                ChallengeSystem.StartChallenge();
                break;
                
            case GameState.TribalCouncil:
                TribalCouncilSystem.PrepareTribalCouncil();
                break;
                
            case GameState.Merge:
                MergeTribes();
                _currentPhase = GamePhase.PostMerge;
                break;
                
            case GameState.FinalTribalCouncil:
                TribalCouncilSystem.PrepareFinalTribalCouncil();
                _currentPhase = GamePhase.Final;
                break;
                
            case GameState.GameOver:
                break;
        }
    }

    public void SelectCharacter(SurvivorData survivor)
    {
        _playerSurvivor = survivor;
        SetGameState(GameState.TribeDivision);
    }

    private void DivideTribes()
    {
        // Determine number of tribes (2 or 3)
        int numberOfTribes = Random.Range(0, 2) == 0 ? 2 : 3;
        
        // Load tribe names
        TextAsset tribeNamesJson = Resources.Load<TextAsset>("Data/TribeNames");
        List<string> tribeNames = new List<string>();
        if (tribeNamesJson != null)
        {
            TribeNameDatabase database = JsonUtility.FromJson<TribeNameDatabase>(tribeNamesJson.text);
            tribeNames = database.tribeNames;
        }
        else
        {
            // Fallback tribe names
            tribeNames = new List<string> { "Apa", "Moto", "Galang", "Tadhana", "Bayon", "Malolo", "Vokai", "Kama", "Lairo", "Yara" };
        }
        
        // Create tribes
        _tribes.Clear();
        Color[] tribeColors = { Color.red, Color.blue, Color.green };
        
        for (int i = 0; i < numberOfTribes; i++)
        {
            TribeData tribe = new TribeData();
            tribe.TribeName = tribeNames[Random.Range(0, tribeNames.Count)];
            tribeNames.Remove(tribe.TribeName); // Remove used name
            tribe.TribeColor = tribeColors[i];
            _tribes.Add(tribe);
        }
        
        // Shuffle all survivors
        List<SurvivorData> shuffledSurvivors = new List<SurvivorData>(_allSurvivors);
        for (int i = 0; i < shuffledSurvivors.Count; i++)
        {
            int randomIndex = Random.Range(i, shuffledSurvivors.Count);
            SurvivorData temp = shuffledSurvivors[i];
            shuffledSurvivors[i] = shuffledSurvivors[randomIndex];
            shuffledSurvivors[randomIndex] = temp;
        }
        
        // Replace one random survivor with player's selected survivor
        int playerIndex = Random.Range(0, shuffledSurvivors.Count);
        shuffledSurvivors[playerIndex] = _playerSurvivor;
        
        // Divide survivors into tribes evenly
        int survivorsPerTribe = _survivorsRemaining / numberOfTribes;
        int remainingSurvivors = _survivorsRemaining % numberOfTribes;
        
        int currentIndex = 0;
        for (int i = 0; i < numberOfTribes; i++)
        {
            int tribeSize = survivorsPerTribe + (i < remainingSurvivors ? 1 : 0);
            for (int j = 0; j < tribeSize; j++)
            {
                _tribes[i].Members.Add(shuffledSurvivors[currentIndex]);
                
                // Track which tribe the player is in
                if (shuffledSurvivors[currentIndex] == _playerSurvivor)
                {
                    _playerTribe = _tribes[i];
                }
                
                currentIndex++;
            }
        }
    }

    private void MergeTribes()
    {
        // Create merged tribe
        _mergedTribe = new TribeData();
        _mergedTribe.TribeName = "Merged";
        _mergedTribe.TribeColor = Color.yellow;
        
        // Add all remaining survivors to merged tribe
        foreach (var tribe in _tribes)
        {
            foreach (var member in tribe.Members)
            {
                _mergedTribe.Members.Add(member);
            }
        }
        
        // Replace tribe list with just the merged tribe
        _tribes.Clear();
        _tribes.Add(_mergedTribe);
        _playerTribe = _mergedTribe;
    }

    public void EliminateSurvivor(SurvivorData survivor)
    {
        // Remove survivor from their tribe
        foreach (var tribe in _tribes)
        {
            if (tribe.Members.Contains(survivor))
            {
                tribe.Members.Remove(survivor);
                break;
            }
        }
        
        // Add to jury if appropriate
        if (_survivorsRemaining <= 10)
        {
            _jury.Add(survivor);
        }
        
        _survivorsRemaining--;
        
        // Check if merge should happen
        if (_survivorsRemaining <= _mergeThreshold && _currentPhase == GamePhase.PreMerge)
        {
            SetGameState(GameState.Merge);
        }
        // Check if final tribal council should happen
        else if (_survivorsRemaining <= _finalThreshold)
        {
            SetGameState(GameState.FinalTribalCouncil);
        }
        // Otherwise continue with the game loop
        else
        {
            _currentDay++;
            SetGameState(GameState.Camp);
        }
    }

    public GameState GetCurrentState()
    {
        return _currentState;
    }

    public GamePhase GetCurrentPhase()
    {
        return _currentPhase;
    }

    public int GetCurrentDay()
    {
        return _currentDay;
    }

    public SurvivorData GetPlayerSurvivor()
    {
        return _playerSurvivor;
    }

    public TribeData GetPlayerTribe()
    {
        return _playerTribe;
    }

    public List<TribeData> GetTribes()
    {
        return _tribes;
    }

    public List<SurvivorData> GetJury()
    {
        return _jury;
    }

    public int GetSurvivorsRemaining()
    {
        return _survivorsRemaining;
    }
    
    [System.Serializable]
    private class SurvivorDatabase
    {
        public List<SurvivorData> survivors;
    }
    
    [System.Serializable]
    private class TribeNameDatabase
    {
        public List<string> tribeNames;
    }
}

File: js/GameManager.js
// Game Manager
class GameManager {
    constructor() {
        // Game state
        this.gameState = "welcome"; // welcome, characterSelection, tribeDivision, camp, challenge, tribalCouncil, merge, finalTribalCouncil, gameOver
        this.gamePhase = "preMerge"; // preMerge, postMerge, final
        this.day = 1;
        this.tribes = [];
        this.playerCharacter = null;
        this.jury = [];
        this.mergeDay = 12;
        
        // Initialize systems
        this.dialogueSystem = new DialogueSystem(this);
        this.energySystem = new EnergySystem(this);
        this.relationshipSystem = new RelationshipSystem(this);
        this.allianceSystem = new AllianceSystem(this);
        this.challengeSystem = new ChallengeSystem(this);
        this.tribalCouncilSystem = new TribalCouncilSystem(this);
        this.idolSystem = new IdolSystem(this);
    }
    
    /**
     * Initialize a new game
     */
    initializeGame() {
        // Reset state
        this.gameState = "welcome";
        this.gamePhase = "preMerge";
        this.day = 1;
        this.tribes = [];
        this.playerCharacter = null;
        this.jury = [];
        
        // Initialize systems
        this.energySystem.initialize();
        this.relationshipSystem.initialize();
        this.allianceSystem.initialize();
        this.challengeSystem.initialize();
        this.tribalCouncilSystem.initialize();
        this.idolSystem.initialize();
        
        // Set up welcome screen
        WelcomeScreen.setup();
    }
    
    /**
     * Start the game
     */
    startGame() {
        this.setGameState("characterSelection");
    }
    
    /**
     * Set the game state and update UI
     * @param {string} newState - The new game state
     */
    setGameState(newState) {
        this.gameState = newState;
        
        // Update UI based on new state
        switch (newState) {
            case "welcome":
                showScreen("welcome-screen");
                WelcomeScreen.setup();
                break;
                
            case "characterSelection":
                showScreen("character-selection-screen");
                CharacterSelectionScreen.setup();
                break;
                
            case "tribeDivision":
                showScreen("tribe-division-screen");
                TribeDivisionScreen.setup();
                break;
                
            case "merge":
                // Process tribe merge when entering this state
                this.mergeTribes();
                showScreen("tribe-division-screen");
                TribeDivisionScreen.setupMerge();
                break;
                
            case "camp":
                this.energySystem.refillEnergy();
                showScreen("camp-screen");
                CampScreen.setup();
                break;
                
            case "challenge":
                showScreen("challenge-screen");
                ChallengeScreen.setup();
                break;
                
            case "tribalCouncil":
                showScreen("tribal-council-screen");
                TribalCouncilScreen.setup();
                break;
                
            case "finalTribalCouncil":
                // Handle final tribal council
                break;
                
            case "gameOver":
                // Handle game over
                this.showGameOverScreen();
                break;
        }
    }
    
    /**
     * Select a character to play as
     * @param {Object} survivor - The selected survivor
     */
    selectCharacter(survivor) {
        // Clone survivor data to avoid reference issues
        this.playerCharacter = deepCopy(survivor);
        this.playerCharacter.isPlayer = true;
        
        // Create tribes
        this.createTribes();
        
        // Proceed to tribe division
        this.setGameState("tribeDivision");
    }
    
    /**
     * Create tribes for the game
     */
    createTribes() {
        // Reset tribes
        this.tribes = [];
        
        // Create two initial tribes
        const tribe1 = {
            tribeName: tribeNames[0],
            tribeColor: tribeColors[0],
            members: [],
            fire: 50,
            water: 50,
            food: 50
        };
        
        const tribe2 = {
            tribeName: tribeNames[1],
            tribeColor: tribeColors[1],
            members: [],
            fire: 50,
            water: 50,
            food: 50
        };
        
        // Add player to tribe 1
        tribe1.members.push(this.playerCharacter);
        
        // Get random NPCs
        const npcSurvivors = [];
        survivorDatabase.survivors.forEach(survivor => {
            npcSurvivors.push(deepCopy(survivor));
        });
        
        // Shuffle the NPCs
        const shuffledNPCs = shuffleArray(npcSurvivors);
        
        // Divide NPCs between tribes
        for (let i = 0; i < 9; i++) {
            if (i < 5) {
                tribe1.members.push(shuffledNPCs[i]);
            } else {
                tribe2.members.push(shuffledNPCs[i]);
            }
        }
        
        // Add tribes to game
        this.tribes.push(tribe1);
        this.tribes.push(tribe2);
        
        // Initialize relationships
        this.tribes.forEach(tribe => {
            this.relationshipSystem.initializeTribeRelationships(tribe);
        });
    }
    
    /**
     * Merge tribes into one
     */
    mergeTribes() {
        if (this.tribes.length <= 1) return;
        
        // Change game phase
        this.gamePhase = "postMerge";
        
        // Create merged tribe
        const mergedTribe = {
            tribeName: tribeNames[2], // Use third tribe name for merged tribe
            tribeColor: tribeColors[2], // Use third color for merged tribe
            members: [],
            fire: 50,
            water: 50,
            food: 50
        };
        
        // Combine members from all tribes
        this.tribes.forEach(tribe => {
            tribe.members.forEach(member => {
                mergedTribe.members.push(member);
            });
        });
        
        // Replace tribes with merged tribe
        this.tribes = [mergedTribe];
        
        // Update relationships for newly merged tribe
        this.relationshipSystem.initializeTribeRelationships(mergedTribe);
    }
    
    /**
     * Get the player's tribe
     * @returns {Object} The player's tribe
     */
    getPlayerTribe() {
        for (const tribe of this.tribes) {
            for (const member of tribe.members) {
                if (member.isPlayer) {
                    return tribe;
                }
            }
        }
        return null;
    }
    
    /**
     * Get the player survivor object
     * @returns {Object} The player survivor
     */
    getPlayerSurvivor() {
        for (const tribe of this.tribes) {
            for (const member of tribe.members) {
                if (member.isPlayer) {
                    return member;
                }
            }
        }
        return null;
    }
    
    /**
     * Get all tribes
     * @returns {Array} All tribes
     */
    getTribes() {
        return this.tribes;
    }
    
    /**
     * Get the current game phase
     * @returns {string} The current game phase
     */
    getGamePhase() {
        return this.gamePhase;
    }
    
    /**
     * Get the current game state
     * @returns {string} The current game state
     */
    getGameState() {
        return this.gameState;
    }
    
    /**
     * Get the current day
     * @returns {number} The current day
     */
    getDay() {
        return this.day;
    }
    
    /**
     * Advance to the next day
     */
    advanceDay() {
        this.day++;
        
        // Clear immunity
        this.clearImmunity();
        
        // Process NPC alliance formations
        this.allianceSystem.processNPCAllianceFormations();
        
        // Process NPCs finding idols
        this.idolSystem.processNPCIdolFinds();
        
        // Check for merge
        this.checkForMerge();
    }
    
    /**
     * Check if tribes should merge
     */
    checkForMerge() {
        if (this.gamePhase === "preMerge" && this.day >= this.mergeDay) {
            this.setGameState("merge");
        }
    }
    
    /**
     * Clear immunity from all survivors
     */
    clearImmunity() {
        this.tribes.forEach(tribe => {
            tribe.members.forEach(member => {
                member.hasImmunity = false;
            });
        });
    }
    
    /**
     * Eliminate a survivor
     * @param {Object} survivor - The survivor to eliminate
     */
    eliminateSurvivor(survivor) {
        if (!survivor) return;
        
        // Find the survivor's tribe
        const tribe = this.tribes.find(t => 
            t.members.some(m => m.name === survivor.name)
        );
        
        if (!tribe) return;
        
        // Remove from tribe
        tribe.members = tribe.members.filter(m => m.name !== survivor.name);
        
        // In post-merge, add to jury
        if (this.gamePhase === "postMerge") {
            this.jury.push(survivor);
        }
        
        // Check if we've reached final 3
        if (this.gamePhase === "postMerge" && tribe.members.length <= 3) {
            this.gamePhase = "final";
        }
        
        // Proceed to camp or final tribal council
        if (this.gamePhase === "final") {
            this.setGameState("finalTribalCouncil");
        } else {
            this.setGameState("camp");
        }
    }
    
    /**
     * Get the jury members
     * @returns {Array} The jury members
     */
    getJury() {
        return this.jury;
    }
    
    /**
     * Save the current game state
     */
    saveGame() {
        const gameData = {
            gameState: this.gameState,
            gamePhase: this.gamePhase,
            day: this.day,
            tribes: this.tribes,
            playerCharacter: this.playerCharacter,
            jury: this.jury
        };
        
        saveGame(gameData);
    }
    
    /**
     * Load a saved game
     */
    loadGame() {
        const savedGame = loadGame();
        
        if (savedGame) {
            this.gameState = savedGame.gameState;
            this.gamePhase = savedGame.gamePhase;
            this.day = savedGame.day;
            this.tribes = savedGame.tribes;
            this.playerCharacter = savedGame.playerCharacter;
            this.jury = savedGame.jury;
            
            // Initialize systems with saved data
            this.energySystem.initialize();
            
            // Update UI based on saved state
            this.setGameState(this.gameState);
            
            return true;
        }
        
        return false;
    }
    
    /**
     * Show game over screen
     */
    showGameOverScreen() {
        const gameOverMessage = "You have been voted off Survivor Island!";
        
        // Display game over message
        this.dialogueSystem.showDialogue(
            gameOverMessage,
            ["Play Again"],
            () => {
                this.dialogueSystem.hideDialogue();
                this.initializeGame();
            }
        );
    }
}